include/openthread/srp_client_service_pool.h

```diff
@@ -0,0 +1,175 @@
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ * @brief
+ *  This file defines the OpenThread SRP (Service Registration Protocol) client item (service/host) pool.
+ */
+
+#ifndef OPENTHREAD_SRP_CLIENT_SERVICE_POOL_H_
+#define OPENTHREAD_SRP_CLIENT_SERVICE_POOL_H_
+
+#include <openthread/dns.h>
+#include <openthread/srp_client.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @addtogroup api-srp
+ *
+ * @brief
+ *   This module includes functions for SRP client service pool.
+ *
+ * @{
+ *
+ * Functions in this module are only available when feature OPENTHREAD_CONFIG_SRP_CLIENT_SERVICE_POOL_ENABLE is enabled.
+ *
+ */
+
+/**
+ * This struct represents a SRP client service pool entry.
+ *
+ */
+typedef struct otSrpClientPoolServiceEntry
+{
+    otSrpClientService mService;  ///< The SRP client service structure.
+    otDnsTxtEntry      mTxtEntry; ///< The SRP client TXT entry.
+} otSrpClientPoolServiceEntry;
+
+/**
+ * This function gets the string buffer to use for SRP client host name.
+ *
+ * @param[in] aInstance   A pointer to the OpenThread instance.
+ * @param[out] aSize      If not NULL, returns the size (number of bytes) of the string buffer.
+ *
+ * @returns A pointer to char buffer to use for SRP client host name.
+ *
+ */
+char *otSrpClientPoolGetHostNameString(otInstance *aInstance, uint16_t *aSize);
+
+/**
+ * This function gets the array of IPv6 address entries to use as SRP client host address list.
+ *
+ * @param[in] aInstance      A pointer to the OpenThread instance.
+ * @param[out] aArrayLength  If not NULL, returns the array length (number of IPv6 address entries in the array).
```

Comment:
I think `aArrayLength` MUST not be `NULL`. Are there chances that the user can/should use the returned address buffer without knowing its length?

Reply:
We can change this and require a non-`NULL` length variable in all the similar getter functions, but I think it may be useful in some cases to let the caller use `NULL`.

The caller may aleady know the length of the array and/or string buffers from having access to the project specific config that it defined when building openthread library (i.e. basically from knowing the value of the `OPENTHREAD_CONFIG_SRP_CLIENT_SERVICE_POOL_MAX_HOST_ADDRSSES` config, etc.).  This situation can happen when OT is integrated into a product where the modules using OT (and its API) have access to the config header file that was provided to OT (e.g., this can happen in SoC style product integration).

Reply:
For cases that the user has access to `OPENTHREAD_CONFIG_SRP_CLIENT_SERVICE_POOL_MAX_HOST_ADDRSSES`, I feel it still not a good practice to use the configuration directly but not get the length along with the returned pointer.

Besides, disallow `NULL` should be able to simplifies the implementation a bit :)

Reply:
I agree it makes the code simpler. I changed this in the latest pushed commit.